src/include/sof/audio/dp_queue.h

```diff
@@ -0,0 +1,114 @@
+/* SPDX-License-Identifier: BSD-3-Clause
+ *
+ * Copyright(c) 2023 Intel Corporation. All rights reserved.
+ *
+ */
+
+#ifndef __SOF_DP_QUEUE_H__
+#define __SOF_DP_QUEUE_H__
+
+#include <sof/audio/sink_api.h>
+#include <sof/audio/source_api.h>
+#include <rtos/bit.h>
+
+/**
+ * DP queue is a lockless circular buffer
+ * providing safe consumer/producer cached operations cross cores
+ *
+ * prerequisites:
+ *  1) incoming and outgoing data rate MUST be the same
+ *  2) Both data consumer and data producer declare max chunk sizes they want to use (IBS/OBS)
+ *
+ * required Buffer size:
+ *	- 2*MAX(IBS,OBS) if the larger of IBS/OBS is multiplication of smaller
+ *	- 3*MAX(IBS,OBS) otherwise
+ *
+ * The queue may work in 2 modes
+ * 1) local mode
+ *    in case both receiver and sender are located on the same core and cache coherency
+ *    does not matter. dp_queue structure is located in cached memory
+ *    In this case DP Queue is a simple ring buffer
+ *
+ * 2) shared mode
+ *    In this case we need to writeback cache when new data arrive and invalidate cache on
+ *    secondary core. dp_queue structure is located in shared memory
+ *
+ *
+ * dpQueue is a lockless consumer/producer safe buffer. It is acheved by having only 2 shared
+ * variables:
+ *  write_offset - can be modified by data producer only
+ *  read_offset - can be modified by data consumer only
+ *
+ *  as 32 bit operations are atomic, it is multi-thread and multi-core save
+ *
+ * There some explanation needed how free_space and available_data are calculated
+ *
+ * number of avail data in circular buffer may be calculated as:
+ *	data_avail = write_offset - read_offset
+ *   and check for wrap around
+ *	if (data_avail < 0) data_avail = buffer_size - data_avail
+ *
+ * The problem is when write_offset == read_offset,
+ * !!! it may mean either that the buffer is empty or the buffer is completely filled !!!
+ *
+ * To solve the above issue having only 2 variables mentioned before:
+ *  - allow both offsets to point from 0 to DOUBLE buffer_size
+ *  - when calculating pointers to data, use: data_bufer[offset % buffer_size]
+ *  - use double buffer size in wrap around check when calculating available data
+ *
+ * And now:
+ *   - write_offset == read_offset
+ *		always means "buffer empty"
+ *   - write_offset == read_offset + buffer_size
+ *		always means "buffer full"
+ */
```

Comment:
the main differences to the current audio_stream.h implementation seem to be:
1. DP Queue doesn't overwrite data - if the buffer is full, writing will fail
2. DP Queue uses the "double buffer size" trick to avoid a shared "buffer full" flag, which allows lockless use
3. The use of cached or uncached pointers for the control structure is tightly connected to the buffer being shared or not
4. It's using the sink / source API

Is that correct? So in principle this buffer implementation could be made generic and used for all audio buffers, but since our current audio-stream buffers have to support overwriting, the producer has to be able to update the read-pointer, so it needs locking

Reply:
_but since our current audio-stream buffers have to support overwriting_ 

I see audio_stream_produce() does support overwriting, but 
 - for LL it makes no point at all - LL must produce a complete portion of data from input and send a complete portion to output. If any of LL fails - it is a glitch anyway
 -   for DP - a module will be run only if it is ready to process, in common case it means "have enough data on all inputs and enough space on all outputs" 

so yes _:DP Queue doesn't overwrite data - if the buffer is full, writing will fail_. In each case - overwrite or drop - data stream is corrupted

_"So in principle this buffer implementation could be made generic and used for all audio buffers,"_
No, it is to heavy to be at each place

There's a need for a couple more queues:

1) simple reusable LL buffer
 For ALL single-core LLs in chain we do need a **single** memory space for sink and a **single** **source** 
 LL produces and consumes **a complete portion of data**, leaving a source buffer empty. Then a next LL is run. So the empty memory space may be easily re-used as a sink for next LL

 2) cross core LL buffer
  for connected cross core LL pipelines. It may be dpQueue, it will just need a smaller buffer size, 1*MAX(ibs,obs) in that case + some tiny details. May be done i,e, by another DP_QUEUE_MODE_ flag

3) linearization dpQueue
the heaviest of all, to be used if a module required a linear data instead of circular, providing internal linearization of data. My idea is that the module may request this somehow (TBD how) and a proper queue is created for it. Note that is applies to DP only. LLs will have linear data in natural way (always start from buffer beginning)